### PR TITLE
Fix and enable scrypt sse2 integration

### PR DESCRIPTION
**Pull request by patricklodder**
Currently the SSE2 scrypt implementation is broken. The underlying SHA256 is working with SSE2 by inheritance from Bitcoin, but notoriously the `xor_salsa8` is always using the generic function. This PR fixes and enables it by:

1. Removing the `printf` calls in the autodetection routine because we use it in more places than just the daemon. I am aware that the current Litecoin implementation [returns a string that tells us what to log](https://github.com/litecoin-project/litecoin/blob/fa0cdba24a0818c4a7206f282c86f0eac6767e72/src/crypto/scrypt.cpp#L317), but I replaced it with a boolean instead, so that, like everything else, the log message is defined in the context the function is called in. I initially played with the idea to make this a little bitmap to be able to indicate whether we are enforcing SSE2 (for x86_64) or not, but left that out, because I think that we should look into a more unified approach to runtime capability detection. [Bitcoin already does something like that for SHA256/x86 in the version we have ported into 1.21](https://github.com/dogecoin/dogecoin/blob/e73ab3f90fcd8bcb20727e2c324744cd4911c82a/src/crypto/sha256.cpp#L562) - but since we have a lot more to consider than just SHA256 and x86, we may want to build a little bit upon that later. I'll do a separate proposal for that.
2. Remove the reliance on passing `-DUSE_SSE2` into `CPPFLAGS` and instead neatly autoconf into `config/bitcoin-config.h`. I kept the `USE_SSE2` flag the same in the source code right now, but this should really be `USE_SCRYPT_SSE2` - there's more than just scrypt using SSE2 and it's confusing. Since the Bitcoin code makes assumptions about SSE2 availability already, I made the build flag `--enable-scrypt-sse2`. Note that it is disabled by default just like all the other intrinsics enhancements, so we can incubate this as an experimental optimization first. I think this is important because frankly, the implementation looks messy after all these years, and we could use some more refactoring on this. If someone wants to do a proposal, be my guest. Otherwise, I'll try to do one myself.
3. Enable sse2 in the x86_64 experimental build.

Test guidance: if you want to test locally, make sure you have a clean build directory, and you can then select this to be included with `./configure --enable-scrypt-sse2`.

### CODE REVIEW

**Review by edtubbs — APPROVED** [review]
Reviewed and tested on i7 with Ubuntu 18.04, bench results below, ACK

bench without `--enable-scrypt-sse2`:

```
Scrypt,4096,0.000231302808970,0.000290848314762,0.000256672152318,488510,614269,542090
```

bench with `--enable-scrypt-sse2`:

```
Scrypt,5120,0.000193746294826,0.000242726877332,0.000206633005291,409191,512624,436408
```

Agreed wrt unified approach for runtime checks (e.g. cpuid.h).

**Comment by victorsk2019** [comment]
I tried testing build from this repo location https://github.com/patricklodder/dogecoin.git, however bench_dogecoin is not producing any "Scrypt" output as in the case with @edtubbs sample.  I'm running Virtual Box Calculate Linux system on a Mac with 4 processors allocated for VB machine.  This is my 'cpuinfo' output:

```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 158
model name      : Intel(R) Core(TM) i5-8500 CPU @ 3.00GHz
stepping        : 10
cpu MHz         : 2999.998
cache size      : 9216 KB
physical id     : 0
siblings        : 4
core id         : 0
cpu cores       : 4
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase avx2 invpcid rdseed clflushopt md_clear flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit srbds
bogomips        : 5999.99
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:

processor       : 1
vendor_id       : GenuineIntel
cpu family      : 6
model           : 158
model name      : Intel(R) Core(TM) i5-8500 CPU @ 3.00GHz
stepping        : 10
cpu MHz         : 2999.998
cache size      : 9216 KB
physical id     : 0
siblings        : 4
core id         : 1
cpu cores       : 4
apicid          : 1
initial apicid  : 1
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase avx2 invpcid rdseed clflushopt md_clear flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit srbds
bogomips        : 5999.99
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:

processor       : 2
vendor_id       : GenuineIntel
cpu family      : 6
model           : 158
model name      : Intel(R) Core(TM) i5-8500 CPU @ 3.00GHz
stepping        : 10
cpu MHz         : 2999.998
cache size      : 9216 KB
physical id     : 0
siblings        : 4
core id         : 2
cpu cores       : 4
apicid          : 2
initial apicid  : 2
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase avx2 invpcid rdseed clflushopt md_clear flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit srbds
bogomips        : 5999.99
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:

processor       : 3
vendor_id       : GenuineIntel
cpu family      : 6
model           : 158
model name      : Intel(R) Core(TM) i5-8500 CPU @ 3.00GHz
stepping        : 10
cpu MHz         : 2999.998
cache size      : 9216 KB
physical id     : 0
siblings        : 4
core id         : 3
cpu cores       : 4
apicid          : 3
initial apicid  : 3
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase avx2 invpcid rdseed clflushopt md_clear flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit srbds
bogomips        : 5999.99
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
```
The **sse2** flag does seem to be supported, however, 'bench_dogecoin' gives the following output:

```
#Benchmark,count,min,max,average,min_cycles,max_cycles,average_cycles
Base58CheckEncode,589824,0.000001843996870,0.000002359971404,0.000001880683057,5529,7076,5639
Base58Decode,1310720,0.000000817506589,0.000000919204467,0.000000832376099,2451,2756,2495
Base58Encode,983040,0.000001003907528,0.000001508917194,0.000001029829824,3010,4524,3087
CCheckQueueSpeed,11264,0.000062945298851,0.000100544886664,0.000088840744204,188728,301482,266388
CCheckQueueSpeedPrevectorJob,768,0.000801952555776,0.001713141798973,0.001302175844709,2404670,5136853,3904574
CCoinsCaching,147456,0.000006937247235,0.000007774273399,0.000007177665894,20801,23311,21522
CoinSelection,416,0.002402402460575,0.002851948142052,0.002470692762962,7203639,8551503,7408368
LockedPool,640,0.001649327576160,0.001934751868248,0.001709026470780,4945500,5801412,5124513
MempoolEviction,49152,0.000020531530026,0.000023669796064,0.000021450624142,61563,70973,64319
RIPEMD160,480,0.002107873558998,0.002324938774109,0.002152112623056,6320508,6971241,6453107
RollingBloom-refresh,1,0.000172000000000,0.000172000000000,0.000172000000000
RollingBloom-refresh,1,0.000093000000000,0.000093000000000,0.000093000000000
RollingBloom-refresh,1,0.000112000000000,0.000112000000000,0.000112000000000
RollingBloom-refresh,1,0.000128000000000,0.000128000000000,0.000128000000000
RollingBloom-refresh,1,0.000094000000000,0.000094000000000,0.000094000000000
RollingBloom-refresh,1,0.000094000000000,0.000094000000000,0.000094000000000
RollingBloom-refresh,1,0.000093000000000,0.000093000000000,0.000093000000000
RollingBloom-refresh,1,0.000094000000000,0.000094000000000,0.000094000000000
RollingBloom-refresh,1,0.000127000000000,0.000127000000000,0.000127000000000
RollingBloom-refresh,1,0.000095000000000,0.000095000000000,0.000095000000000
RollingBloom-refresh,1,0.000094000000000,0.000094000000000,0.000094000000000
RollingBloom-refresh,1,0.000093000000000,0.000093000000000,0.000093000000000
RollingBloom-refresh,1,0.000093000000000,0.000093000000000,0.000093000000000
RollingBloom-refresh,1,0.000093000000000,0.000093000000000,0.000093000000000
RollingBloom-refresh,1,0.000101000000000,0.000101000000000,0.000101000000000
RollingBloom-refresh,1,0.000114000000000,0.000114000000000,0.000114000000000
RollingBloom-refresh,1,0.000093000000000,0.000093000000000,0.000093000000000
RollingBloom-refresh,1,0.000128000000000,0.000128000000000,0.000128000000000
RollingBloom-refresh,1,0.000093000000000,0.000093000000000,0.000093000000000
RollingBloom-refresh,1,0.000094000000000,0.000094000000000,0.000094000000000
RollingBloom-refresh,1,0.000110000000000,0.000110000000000,0.000110000000000
RollingBloom-refresh,1,0.000095000000000,0.000095000000000,0.000095000000000
RollingBloom-refresh,1,0.000093000000000,0.000093000000000,0.000093000000000
RollingBloom-refresh,1,0.000094000000000,0.000094000000000,0.000094000000000
RollingBloom,1441792,0.000000677093340,0.000000724747224,0.000000707465336,2030,2173,2121
SHA1,640,0.001576576381922,0.001848876476288,0.001602821797132,4727356,5543873,4806057
SHA256,256,0.003925681114197,0.004912495613098,0.004036133177578,11771161,14730191,12102346
SHA256_32b,4,0.280641555786133,0.287926435470581,0.284283995628357,841503516,863348190,852425853
SHA512,416,0.002386622130871,0.002814576029778,0.002494211953420,7156331,8439600,7478896
SipHash_32b,44,0.023208260536194,0.026489019393921,0.023895659229972,69590157,79427430,71651172
Sleep100ms,10,0.105825066566467,0.125981450080872,0.117265605926514,317316157,377755903,351621117
Trig,67108864,0.000000014025261,0.000000015463826,0.000000014995472,42,46,44
```
I most definitely use '--enable-scrypt-sse2' as a configure option:
```
 ./configure --prefix=/home/me/dogecoind --enable-cxx --with-incompatible-bdb --with-gui=qt5 --with-qt-incdir=/usr/include/qt5 --enable-scrypt-sse2
```
And I also tried adding: CPPFLAGS="-DUSE_SCRYPT_SSE2" as a configure option, but the resulting build is like scrypt_sse2 not even there 🤷🏻‍♂️  Maybe it has something to do with the fact that I am running VirtualBox, but as mentioned above CPU info shows sse2 support should work?

**Comment by victorsk2019** [comment]
I've been chewing on SSE2 scrypt support because I was curious to see what bench test will be like with SSE2 scrypt turned on.  Knowing to see if additional flags, like SSE2 work is important to me when packaging Dogecoin for Gentoo which allows to easily trigger on/off additional flags during installation.

I was able to modify 'dogecoin-1.14.6-dev' build with Scrypt SSE2 support, here are results:

Scrypt without SSE2:
```
Scrypt,4096,0.000253062695265,0.000311465002596,0.000260479282588,758799,933924,781046
```
```
scrypt: using scrypt-sse2 as built.
Scrypt,5632,0.000176656059921,0.000187829136848,0.000179460399191,529701,563226,538112
```
I was using the following configure option:
```
./configure --prefix=/home/me/dogecoind --enable-cxx --with-incompatible-bdb --enable-scrypt-sse2 CPPFLAGS="-DUSE_SSE2" --with-gui=no
```
This is because for "USE_SSE2" flag there is scrypt_detect_sse2() function in 'init.cpp' from crypt/scrypt.h which triggers scrypt_sse2 support.  I then added include in init.cpp:

```
#ifdef USE_SSE2
#include "crypto/scrypt.h"
#endif
```
To fix error "scrypt_detect_sse2()" was not declared in this scope.  Also, I moved functions xor_salsa8_sse2() and scrypt_1024_1_1_256_sp_sse2() from scrypt-sse2.cpp to scrypt.cpp and added header files <cstdio> and <emmintrin.h> in scrypt.cpp from scrypt-sse2.cpp.  So maybe scrypt-sse2.cpp is not necessary anymore as scrypt.cpp can also handle sse2 operations?  I also added standard namespace for printf() to fix printf() was not declared in this scope error in scrypt.cpp: 
```
#if defined(USE_SSE2_ALWAYS)
    std::printf("scrypt: using scrypt-sse2 as built.\n");
#else
```
I document these modifications here to show what steps I took to make Scrypt with SSE2 support work.  Not sure if I should make a PR for this?

**Comment by patricklodder** [comment]
> I document these modifications here to show what steps I took to make Scrypt with SSE2 support work. Not sure if I should make a PR for this?

This PR already fixes this. Which branch from my repo did you build exactly?

**Comment by victorsk2019** [comment]
> This PR already fixes this. Which branch from my repo did you build exactly?

Well, I used 1.14.6-dev from your repository: git clone https://github.com/patricklodder/dogecoin.git

```
git branch --show-current
1.14.6-dev
```
but even Scrypt wasn't working there, which got me a bit confused how others were able to get Scrypt SSE2 working, so I went ahead working on our main 1.14.6 developer build, which had Scrypt but not SSE2 working and I modified it to work with SSE2 (as per my notes above).  Oh well, I familiarized with source code more.  Which branch should I have used?

**Comment by patricklodder** [comment]
> Which branch should I have used?

`1.14.6-fix-scrypt-sse2` per this pull request

**Comment by victorsk2019** [comment]
> > Which branch should I have used?
> 
> `1.14.6-fix-scrypt-sse2` per this pull request

Indeed!  This version works great with '--enable-scrypt-sse2'! 👍
```Scrypt,5632,0.000176988076419,0.000190827995539,0.000179957894778,530960,572484,539873```
And I also understand this implementation better and can compare it with mine 😎

**Comment by victorsk2019** [comment]
> should really be USE_SCRYPT_SSE2

Hi,

I've modified 1.14.6-fix-scrypt-sse2 with "USE_SCRYPT_SSE2" https://github.com/victorsk2019/dogecoin-1.14.6-dev.  Maybe we should have main branch with --enable-scrypt-sse2 support, it works great! 🚀

**Comment by patricklodder** [comment]
@victorsk2019 cool! thanks! I'd suggest to PR that after this one gets reviewed (fixed where needed) and merged.

I myself have other works that I want to put on top of this too: the generic runtime capability detection, so that we also can enable at the very least further maturity to the ARM intrinsics @edtubbs worked on and maybe get some of the intrinsics work out of experimental, either with this release or the next.

**Comment by chromatic** [comment]
I think there's a `Makefile` problem expressing itself with parallel builds. I built this with `make -j9` and the first run gave this error (some output snipped):

```
  CXXLD    dogecoind
  CXXLD    test/test_dogecoin
  CXXLD    bench/bench_dogecoin
  CXXLD    test/test_dogecoin_fuzzy
  OBJCXXLD qt/dogecoin-qt
  CXXLD    qt/test/test_dogecoin-qt
  CXXLD    libdogecoinconsensus.la
/usr/bin/ld: libdogecoin_server.a(libdogecoin_server_a-init.o): in function `AppInitMain(boost::thread_group&, CScheduler&)':
/home/chromatic/dev/dogecoin/src/init.cpp:1235: undefined reference to `scrypt_detect_sse2()'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:4060: dogecoind] Error 1
make[2]: *** Waiting for unfinished jobs....
/usr/bin/ld: bench/bench_dogecoin-scrypt.o: in function `Scrypt(benchmark::State&)':
/home/chromatic/dev/dogecoin/src/bench/scrypt.cpp:19: undefined reference to `scrypt_detect_sse2()'
/usr/bin/ld: libdogecoin_server.a(libdogecoin_server_a-init.o): in function `AppInitMain(boost::thread_group&, CScheduler&)':
/home/chromatic/dev/dogecoin/src/init.cpp:1235: undefined reference to `scrypt_detect_sse2()'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:4048: bench/bench_dogecoin] Error 1
/usr/bin/ld: libdogecoin_server.a(libdogecoin_server_a-init.o): in function `AppInitMain(boost::thread_group&, CScheduler&)':
/home/chromatic/dev/dogecoin/src/init.cpp:1235: undefined reference to `scrypt_detect_sse2()'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:4106: qt/test/test_dogecoin-qt] Error 1
/usr/bin/ld: test/test_dogecoin-scrypt_tests.o: in function `scrypt_tests::scrypt_hashtest::test_method()':
/home/chromatic/dev/dogecoin/src/test/scrypt_tests.cpp:17: undefined reference to `scrypt_detect_sse2()'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:4250: test/test_dogecoin] Error 1
/usr/bin/ld: libdogecoin_server.a(libdogecoin_server_a-init.o): in function `AppInitMain(boost::thread_group&, CScheduler&)':
/home/chromatic/dev/dogecoin/src/init.cpp:1235: undefined reference to `scrypt_detect_sse2()'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:4078: qt/dogecoin-qt] Error 1
make[2]: Leaving directory '/home/chromatic/dev/dogecoin/src'
make[1]: *** [Makefile:9845: all-recursive] Error 1
make[1]: Leaving directory '/home/chromatic/dev/dogecoin/src'
make: *** [Makefile:691: all-recursive] Error 1
```

Then I ran `make -j9 check` and that finished off the standard target before moving to the `check` target:

```
Making check in src
make[1]: Entering directory '/home/chromatic/dev/dogecoin/src'
make[2]: Entering directory '/home/chromatic/dev/dogecoin/src'
make[3]: Entering directory '/home/chromatic/dev/dogecoin'
make[3]: Leaving directory '/home/chromatic/dev/dogecoin'
  CXX      crypto/libdogecoin_crypto_a-scrypt.o
  CXX      crypto/libdogecoin_crypto_a-scrypt-sse2.o
  CXX      crypto/libdogecoinconsensus_la-scrypt.lo
  CXX      crypto/libdogecoinconsensus_la-scrypt-sse2.lo
  AR       crypto/libdogecoin_crypto.a
/usr/bin/ar: `u' modifier ignored since `D' is the default (see `U')
  CXXLD    dogecoind
  CXXLD    test/test_dogecoin
  CXXLD    bench/bench_dogecoin
  OBJCXXLD qt/dogecoin-qt
  CXXLD    qt/test/test_dogecoin-qt
  CXXLD    dogecoin-cli
  CXXLD    dogecoin-tx
  CXXLD    test/test_dogecoin_fuzzy
  CXXLD    libdogecoinconsensus.la
make  check-TESTS check-local
```

Otherwise, the `make check` and RPC tests all pass on 64-bit x86 Linux. I'll review the diff now.

**Comment by chromatic** [comment]
If I add a dependency on `crypto/scrypt.h` to the `libdogecoin_server_a_SOURCES` automake target, I can get the parallel build working correctly. That seems like a minimal reasonable change.

**Comment by patricklodder** [comment]
@chromatic: I can't reproduce on a clean build - did you build this on a clean dir or one with existing artifacts? If the latter:

- what was the previous commit you built?
- what did you pass to `configure` in the previous build?
- what did you pass to `configure` to run this?

I'll try to repro and identify the issue - if it's a race condition then I need to find some way to repro.

**Comment by chromatic** [comment]
I built this clean, using `./configure --enable-scrypt-sse2`.

**Comment by patricklodder** [comment]
🤔 what distro / version?

**Comment by chromatic** [comment]
I tested on Ubuntu 20.04 LTS, with gcc 9.4.0-1ubuntu1~20.04.1, GNU Make 4.2.1, and GNU Autoconf 2.69.

I have no concerns with the code, so as long as we can figure out what went weird on my system this PR will get my blessing.

**Comment by patricklodder** [comment]
I don't mind adding a pointer to the header, but I need to understand why this is not an issue on the 21.04 I tested this on a week ago, or the 22.04 that I just tested it on. I'll run this on 20.04 and see if I can get the same result.

**Comment by chromatic** [comment]
The only reliable reproduction I've found so far is running `./autogen.sh` on a different branch, switching to this branch, and re-running `./configure`, which definitely does the wrong thing for understandable reasons.

If I were writing this code, I'd have added the header dependency to avoid other problems, but that wouldn't explain the behavior I've seen. If you can't reproduce it in any other fashion @patricklodder, then let's not let it block the PR.

We should have plenty of time in CI and other testing on the branch to trigger any lingering race conditions, if they do indeed exist.

**Comment by patricklodder** [comment]
I couldn't reproduce it either, so I was just writing a Dockerfile for not running `./autogen.sh` because I think the issue is not running `make distclean` when configure.ac is introducing a new variable...

I'm fearful of adding a header from `LIBDOGECOIN_CRYPTO` to `LIBDOGECOIN_SERVER`: we kill modularity that way and we're not doing it anywhere else either. I think you'll have the same finding when introducing a new build rule for any of the library build targets.

**Comment by chromatic** [comment]
I haven't been able to reproduce my error with:

```
make distclean && ./autogen.sh && ./configure --enable-scrypt-sse2 && make -j9 check
```

Given that, and the point about reducing coupling between targets we've previously kept separate, I'm comfortable considering the error I posted an error on my side, and nothing broadly worrying.

File: src/crypto/scrypt.cpp
Comment (chromatic) [inline src/crypto/scrypt.cpp:251]: This line looks redundant, unless there's a chance runtime CPU feature detection can change. I don't mind keeping it so that both branches of the conditional are parallel, but properties of the running system could change as a side effect of calling this function. Is there a measurable risk to consider here?

I know we don't want to change line 223, because that'll break the side effect property if this function takes the branch on line 227.

Or (my slight preference), do we want to skip this question until we get runtime intrinsic detection working reliably?

Comment (patricklodder) [reply src/crypto/scrypt.cpp:251]: In 3a1fd97f, I tried to not change any of the code that was working and do as little refactoring as possible. I agree with your assessment and I agree also that the best way to address it is by implementing the generic runtime capability detection rather than fixing this now - we're likely to remove this completely.

Comment (chromatic) [reply src/crypto/scrypt.cpp:251]: We've just now documented our intent to rework this code in the near future, and that's good enough for me.

**Review by chromatic — APPROVED** [review]
ACK. Tested successfully on 64-bit x86 Linux, and all of my code questions are addressed by our plan to rework this code in the near future.

**Pull request merged**
